File: clojure/beer-song/src/beer_song.clj
(ns beer-song
  (:require [clojure.string :as str]))

(defn x-bottles [num]
  (condp = num
    0 "no more bottles"
    1 "1 bottle"
    (str num " bottles")))

(defn x-bottles-caps [num]
  (str/capitalize (x-bottles num)))

(defn action [num]
  (condp = num
    0 "Go to the store and buy some more"
    1 "Take it down and pass it around"
    "Take one down and pass it around"))

(defn verse [num]
  (format "%s of beer on the wall, %s of beer.\n%s, %s of beer on the wall.\n"
          (x-bottles-caps num)
          (x-bottles num)
          (action num)
          (x-bottles (mod (dec num) 100))))

(defn sing
  ([start] (sing start 0))
  ([start end]
   (str/join "\n"
             (for [i (reverse (range end (inc start)))]
               (verse i)))))

File: scala/beer-song/src/main/scala/BeerSong.scala
object BeerSong {

  def recite(nrBeers: Int, nrVerses: Int): String = {
    (nrBeers to 0 by -1)
      .map {
        case beer if beer == 0 =>
          s"No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n"
        case beer if beer == 1 =>
          s"1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n"
        case beer if beer == 2 =>
          s"2 bottles of beer on the wall, 2 bottles of beer.\nTake one down and pass it around, 1 bottle of beer on the wall.\n"
        case beer =>
          s"$beer bottles of beer on the wall, $beer bottles of beer.\nTake one down and pass it around, ${beer - 1} bottles of beer on the wall.\n"
      }
      .take(nrVerses)
      .mkString("\n")
  }
}
